src/server/mod.rs

```diff
@@ -53,11 +63,230 @@ where
+                        Err(err) if err.is_disconnected() => (true, Some(io_upgrade_fut)),
+                        // the channel should never be full without try_send
+                        Err(err) => unreachable!("Unexpected channel err {}", err),
+                    }
+                }
+                res = io_upgrade_fut => {
+                    match res {
+                        Ok(()) => (true, None),
+                        Err(err) if err.is_disconnected() => (false, Some(io_primary_fut)),
+                        Err(err) => unreachable!("Unexpected channel err {}", err),
+                    }
+                }
+                _ = io_end_fut => {
+                    return
+                }
+            };
+            
+            if is_read_failure {
+                return
+            }
+            
+            if let Some(fut) = unfinished_send {
+                match fut.await {
+                    Ok(()) => (),
+                    Err(err) if err.is_disconnected() => return,
+                    Err(err) => unreachable!("Unexpected channel err {}", err),
+                };
+            }
+            
+            trace!("Reader reading to Body::io {}", is_io_body);
+            let mut io_send = if is_io_body {
+                io_upgrade_send
+            } else {
+                io_primary_send
+            };
+            
+            let mut buf = Vec::new();
+            buf.resize(8 * 1024, 0);
+            let buf = buf.into_boxed_slice();
+            let mut buf_ready = Arc::new(buf);
+            
+            let mut buf = Vec::new();
+            buf.resize(8 * 1024, 0);
+            let buf = buf.into_boxed_slice();
+            let mut buf_pending = Arc::new(buf);
+            
+            let mut read_fut = _io.read(Arc::get_mut(&mut buf_pending).unwrap()).fuse();
+            
+            loop {
+                let res = select! {
+                    res = read_fut => {
+                        std::mem::swap(&mut buf_pending, &mut buf_ready);
+                        
+                        read_fut = _io.read(Arc::get_mut(&mut buf_pending).unwrap()).fuse();
+                        res
+                    }
+                    _ = io_end_fut => {
+                        break
+                    }
+                };
+                
+                n = match res {
+                    Ok(n) => {
+                        n
+                    }
+                    Err(err) => {
+                        match io_send.send(Err(err)).await {
+                            Ok(_) => (),
+                            Err(err) if err.is_disconnected() => (),
+                            Err(err) => unreachable!("Unexpected channel err {}", err),
+                        }
+                        
+                        return
+                    }
+                };
+                
+                match io_send.send(Ok((buf_ready.clone(), n))).await {
```

Comment:
I can't find the receiver end for this but do we need to send if `n == 0`?

Reply:
It's to signal that it's done. io_send is set to one of the two senders on line 224 https://github.com/http-rs/async-h1/pull/134/files/d5585c722124f3d819c100d662016125dffff0cb#diff-afb1712a59f738bc3dfe39004cb5e4b9R224